diff --git a/include/private/android_filesystem_config.h b/include/private/android_filesystem_config.h
index 16efe58..f7d2e88 100644
--- a/include/private/android_filesystem_config.h
+++ b/include/private/android_filesystem_config.h
@@ -99,6 +99,8 @@
 #define AID_MOT_DLNA      9011  /* DLNA native */
 #endif // MOTOROLA_UIDS
 
+#define AID_NVRAM         9997  /* MediaTek NVRAM user */
+
 #define AID_MISC          9998  /* access to misc storage */
 #define AID_NOBODY        9999
 
@@ -175,6 +177,7 @@ static const struct android_id_info android_ids[] = {
     { "mot_caif",  AID_MOT_CAIF, },
     { "mot_dlna",  AID_MOT_DLNA, },
 #endif
+    { "nvram",     AID_NVRAM, },
     { "misc",      AID_MISC, },
     { "nobody",    AID_NOBODY, },
     { "theme_man", AID_THEMEMAN },
